### PR TITLE
Add cats module

### PR DESCRIPTION
**Pull request by lvicentesanchez**
With the new modular approach, an `argonaut-cats` module can be easily added. This is still a WIP but I would like @seanparsons to review it before continuing. 
As `cats` has no lenses support, this module has less features than the `scalaz` module.

### CODE REVIEW

**Comment by seanparsons** [comment]
Looks pretty good, does cats have the same sort of typeclass testing features that are in whichever scalaz subproject?

**Comment by lvicentesanchez** [comment]
@seanparsons yes, there is a cats-law project that help you check that your instances obey the laws

File: argonaut-cats/src/main/scala/argonaut/EncodeJsonCats.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import cats.functor.Contravariant
+
+object EncodeJsonCats extends EncodeJsonCatss {
+}
+
+trait EncodeJsonCatss {
+  def fromFoldable[F[_], A](implicit A: EncodeJson[A], F: Foldable[F]): EncodeJson[F[A]] =
+    EncodeJson(fa => jArray(F.foldLeft(fa, Nil: List[Json])((list, a) => A.encode(a) :: list).reverse))
+
+  implicit def DisjunctionEncodeJson[A, B](implicit ea: EncodeJson[A], eb: EncodeJson[B]): EncodeJson[Xor[A, B]] =
```
Comment (adelbertc) [inline argonaut-cats/src/main/scala/argonaut/EncodeJsonCats.scala]: `XorEncodeJson` ?

File: argonaut-cats/src/main/scala/argonaut/EncodeJsonCats.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    def contramap[A, B](r: EncodeJson[A])(f: B => A) = r contramap f
+  }
+
+  implicit def NonEmptyListEncodeJson[A: EncodeJson]: EncodeJson[NonEmptyList[A]] =
+    fromFoldable[NonEmptyList, A]
+
+  implicit def StreamingEncodeJson[A: EncodeJson]: EncodeJson[Streaming[A]] =
+    fromFoldable[Streaming, A]
+
+  implicit def ValidationEncodeJson[E, A](implicit ea: EncodeJson[E], eb: EncodeJson[A]): EncodeJson[Validated[E, A]] =
```
Comment (adelbertc) [inline argonaut-cats/src/main/scala/argonaut/EncodeJsonCats.scala]: `ValidatedEncodeJson`

File: argonaut-cats/src/main/scala/argonaut/DecodeJsonCats.scala [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+object DecodeJsonCats extends DecodeJsonCatss {
+}
+
+trait DecodeJsonCatss {
+  implicit def XorDecodeJson[A: DecodeJson, B: DecodeJson](implicit DE: DecodeJson[Either[A, B]]): DecodeJson[Xor[A, B]] = {
+    DE.map(Xor.fromEither(_))
+  }
+
+  implicit def ValidationDecodeJson[A, B](implicit DA: DecodeJson[A], DB: DecodeJson[B]): DecodeJson[Validated[A, B]] = {
```
Comment (adelbertc) [inline argonaut-cats/src/main/scala/argonaut/DecodeJsonCats.scala]: `ValidatedEncodeJson`

**Comment by lvicentesanchez** [comment]
@adelbertc Thank you for pointing that out. I need to go over the PR again and fix a few naming issues and add some missing instances. Thank you for pointing that out :)

**Comment by lvicentesanchez** [comment]
I have just done some work to keep naming of instances consistent. I have added codec for most data structures defined in cats; the only one missing is `Ior` the equivalent to scalaz `\&/`

I would add tests for the monad and contravariant instances I have defined and... I think that should be it as a first version of the argonaut-cats module.

**Comment by lvicentesanchez** [comment]
I think I have reached feature parity with the scalaz module (if you don't count lenses), so I don't think this is  WIP any more.

**Comment by seanparsons** [comment]
2.10.X builds are le broken because of Option.contains... :(

**Comment by lvicentesanchez** [comment]
@seanparsons >.< I would not trust Intellij suggestions anymore

**Pull request merged**